### ISSUE TITLE
Detail about testers.GlobalEmbeddingSpaceTester

### ISSUE THREAD

**Issue by brianw0924** [KevinMusgrave/pytorch-metric-learning#551]
Hi, I'm wondering if there are any detail introduction or document for testers.GlobalEmbeddingSpaceTester

I have several questions:

1. If my training dataset and testing dataset have different num_classes. How would tester know the num_classes of val_dataset when applying k-nn? I guess the trainer will get num_classes from argument "labels" passed to Sampler so that TrainWithClassifier will know the num_classes but tester don't have it.
 
2. How does the tester.GlobalEmbeddingSpaceTester know my models is trunk only or trunk + embedder? I saw the example code, it just simply pass the hook to trainer. Will the trainer pass the model to tester, too?

Thank you

**Comment by KevinMusgrave** [comment]
> Hi, I'm wondering if there are any detail introduction or document for testers.GlobalEmbeddingSpaceTester

Relevant docs:

- [Testers](https://kevinmusgrave.github.io/pytorch-metric-learning/testers/)
- [Accuracy calculation](https://kevinmusgrave.github.io/pytorch-metric-learning/accuracy_calculation/)


> 1. If my training dataset and testing dataset have different num_classes. How would tester know the num_classes of val_dataset when applying k-nn? I guess the trainer will get num_classes from argument "labels" passed to Sampler so that TrainWithClassifier will know the num_classes but tester don't have it.

Both training and val datasets are assumed to return labels in their `__getitem__` method. As long as this is the case, the tester has access to the labels of the val dataset. If this isn't the case, you would know because the code will almost certainly crash.

> 2. How does the tester.GlobalEmbeddingSpaceTester know my models is trunk only or trunk + embedder? I saw the example code, it just simply pass the hook to trainer. Will the trainer pass the model to tester, too?

The hook has access to the tester, and it passes your model to the tester. If you don't pass in an embedder, then internally, the embedder defaults to an "Identity" function, which does nothing (it just returns what you pass into it).

**Comment by brianw0924** [comment]
> Both training and val datasets are assumed to return labels in their `__getitem__` method. As long as this is the case, the tester has access to the labels of the val dataset. If this isn't the case, you would know because the code will almost certainly crash.

I understand that the dataset __getitem__ will return label, and my __getitem__ return (data, label)

but the tester is not implement by me (I simply call the trainer.train and pass the end_of_epoch argument). How would the tester know the num_classes if I don't specify it or pass the argument to tester?

**Comment by KevinMusgrave** [comment]
> but the tester is not implement by me (I simply call the trainer.train and pass the end_of_epoch argument). How would the tester know the num_classes if I don't specify it or pass the argument to tester? or did you assume that all dataset's `__getitem__` returns in the format (data, label)?

Yes it assumes that any dataset passed in will return (data, label). So when it gathers all the embeddings (by calling `model(data)`, it also gathers all the labels.

**Comment by brianw0924** [comment]
Sorry. I have another question about samplers.MPerClassSampler

if num_classes=10, m=4, batch_size=32. Does it mean that each iteration will only return 32/4=8 classes? (2 classes will not be selected in this iteration)

**Comment by KevinMusgrave** [comment]
> if num_classes=10, m=4, batch_size=32. Does it mean that each iteration will only return 32/4=8 classes? (2 classes will not be selected in this iteration)

That's correct
